### PR TITLE
DEPS: Make pytz an optional dependency

### LINKED ISSUE #34916: API: Default to stdlib timezone objects instead of pytz
i.e. when we get `pd.Timestamp.now("UTC")` we should return a Timestamp with `datetime.timezone.utc` rather than `pytz.UTC`.  Similarly when we parse an ISO8601 datetime we should use tzinfo of `timezone(timedelta(seconds=val*60))` instead of `pytz.FixedOffset(val)`

This isn't _that_ hard to implement, but doing it breaks a couple dozen tests where we are currently checking for pytz objects.  This would technically be an API change, so putting it up for discussion before implementing it.

### PR DESCRIPTION
**Pull request by mroeschke**
- [ ] closes #34916 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This makes pandas return zoneinfo.ZoneInfo objects by default from string inputs

### CODE REVIEW

File: pandas/_libs/tslibs/timezones.pyx [hunk 1]
```diff
@@ -45,7 +38,7 @@ from pandas._libs.tslibs.util cimport (
 
 cdef int64_t NPY_NAT = get_nat()
 cdef tzinfo utc_stdlib = timezone.utc
-cdef tzinfo utc_pytz = pytz.utc
+cdef object utc_pytz = pytz.UTC if pytz else None
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timezones.pyx:41]: cant we keep this as a tzinfo object?

File: pandas/_libs/tslibs/timezones.pyx [hunk 2]
```diff
@@ -74,18 +67,16 @@ cdef bint is_utc_zoneinfo(tzinfo tz):
...
         or is_utc_zoneinfo(tz)
+        or (utc_pytz is not None and tz is utc_pytz)
     )
 
 
 cdef bint is_zoneinfo(tzinfo tz):
-    if ZoneInfo is None:
-        return False
-    return isinstance(tz, ZoneInfo)
+    return isinstance(tz, zoneinfo.ZoneInfo)
```
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/_libs/tslibs/timezones.pyx]: Sure thing. Added

File: pandas/_libs/tslibs/timezones.pyx (hunk 1 above)
Comment (mroeschke) [reply pandas/_libs/tslibs/timezones.pyx]: Ah I had thought cython wouldn't like tzinfo since this could be `None`, but was able to keep as tzinfo

**Comment by mroeschke** [comment]
Will merge at the end of the week if there are no further comments

File: doc/source/whatsnew/v3.0.0.rst [hunk 3]
```diff
@@ -230,6 +232,37 @@ Optional libraries below the lowest tested version may still work, but are not c
...
+    <DstTzInfo 'US/Pacific' LMT-1 day, 16:07:00 STD>
+
+*New behavior:*
+
+.. ipython:: python
+
+    ts = pd.Timestamp(2024, 1, 1).tz_localize("US/Pacific")
+    ts.tz
+
+``pytz`` timezone objects are still supported when passed directly, but they will no longer be returned by default
```
Comment (baryluk) [inline doc/source/whatsnew/v3.0.0.rst:258]: > ``pytz`` timezone objects are still supported when passed directly, but they will no longer be returned by default

This is a bit confusing. It suggests, that pd.Timestamp and others can be used to make it do return `pytz` objects.

**Pull request merged**